---
title: "Supervised Penalized Canonical Correlation Analysis with the spCCA package"
package: spCCA
output:
  BiocStyle::html_document:
    includes:
      in_header: spcca.bioschemas.html
author:
- name: Andrea Thum
  email: andrea.thum@informatik.uni-halle.de
- name: Steffen Neumann
  email: sneumann@ipb-halle.de
abstract: |
  The supervised penalized canonical correlation approach (spCCA) is an 
   extension of the penalized CCA, where the experimental design is used 
   as a third data set (making the analysis supervised), and the correlation 
   of the biological data sets with the design data set is maximized to find 
   interpretable and meaningful canonical variables. This vignette shows 
   the spCCA analysis on a data set of Arabidopsis thaliana with gene expression 
   and metabolite intensity measurements, resulting in eight significant canonical
   variables. 
vignette: |
  %\VignetteIndexEntry{spCCA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: lit.bib
csl: biomed-central.csl
---

```{r LibraryPreload, echo=FALSE, results='hide'}
library(spCCA)
library(knitr)
library(kableExtra)
```

# Introduction

See [@Hotel36].

# Datasets

We reduced the data sets by excluding features with low variance 
(threshold chosen σ < 1 for genes and σ < 0.4 for metabolites), 
resulting in a 72 × 252 LC/MS signal intensity matrix (X), 
and a 72 × 1277 gene expression matrix (Y).

```{r}
MTBLS18_path <- system.file("arXiv.1405.1534", "MTBLS18.csv", 
                            package = "spCCA")
EMTAB3287_path <- system.file("arXiv.1405.1534", "E-MTAB-3287.csv", 
                                package = "spCCA")
SAMEG179892_path <- system.file("arXiv.1405.1534", "SAMEG179892.csv", 
                                    package = "spCCA")

X <- as.matrix(read.csv(MTBLS18_path, row.names = 1))
Y <- as.matrix(read.csv(EMTAB3287_path, row.names = 1))
Z <- as.matrix(read.csv(SAMEG179892_path, row.names = 1))

```

## Data visualisation

Show a graphical representation of the experimental design factors. 
Then, check the distribution of values in the biological data sets. 

TODO: double check x position of the samples and actual factor levels 
for the experiments/samples.

```{r}

plot(Z[,"DayNight"]*0.9, main="Experimental Design Factors", type="s", 
     xlab="Sample", ylab="",
     ylim=c(0,ncol(Z)), xaxt="n", yaxt="n")
axis(1, at=seq(0, along=rownames(X)), labels=rownames(X), 
     las=2, cex.axis=1/(nrow(Z)/25))
axis(2, at=seq(0, along=colnames(Z)), labels=colnames(Z), 
     las=2, cex.axis=0.6)

abline(v=seq(6.5, along=rownames(X), by=6), col="grey", lty=2)
abline(v=seq(12.5, along=rownames(X), by=12), col="grey", lty=1)

i<-0
dummy <- sapply(colnames(Z), function(f) { lines(Z[,f]*0.9+i, col=i, type="s"); i<<-i+1 })

hist(X, freq=FALSE, main="Metabolites", xlab="Intensity", ylab="Density")
lines(density(X), col="blue")

hist(Y, freq=FALSE, main="Transcripts", xlab="Intensity", ylab="Density")
lines(density(Y), col="blue")



```

# spCCA


Unsure how to perform the "ten-fold repeated hold-out sampling" mentioned in the paper. 


```{r getCCA3, cache=TRUE}

# Define the grid size based on paper
end  <- c(16, 21, 31)
step <- c( 1,  1,  1)

# Define the grid size based on manpage
end  <- c(1.0,  0.2,  2.0)
step <- c(0.05, 0.01, 0.1)

CCA3 <- getCCA3(X=X, Y=Y, Z=Z, 
                numCV=6,
                end=end, step=step,
                n.r = 5, ## number of repeats, increase to 10 when parameters are stable
                max.counter.test = 5 ## number of random start vectors for iteration, increase to 10-20 when parameters are stable
                )  

#CCA3 <- getCCA3(X=X, Y=Y, Z=Z, numCV=9, end=end, step=step)
```

The higher the lambdas, the more sparse the canonical variables will be.

The parameters for the lambda grid search depend on the number of features (rows) 
in the data matrices. 

The end value is inverse proportional to the number of features.
For the metabolomics data with n=252 rows, an end value of 1.0 could be a good value.
For the transcriptomics data with n=1277 rows, an end value of 0.2 could be a good value.
For the design matrix with n=11 design factors, an end value of 2.0 could be a good value.


The step size is inverse proportional to the number of features.
For the metabolomics data with n=252 rows, a step size of 0.05 could be a good value.
For the transcriptomics data with n=1277 rows, a step size of 0.01 or slightly 
smaller is a good value.
For the design matrix with n=11 design factors, a step size of 0.1 could be a good value.


step size 0.01 is a good number for sets with 1000 features

```{r}
  plotCCA(CCA3, X, Y, filename = NULL, legend=c("metabolites", "genes", "design"))
```

## Details of canonical variables

```{r}
numCV <- 1
topN <- 1:10

kables(list(
kable(CCA3$cc3.weight.z[order(abs(CCA3$cc3.weight.z[,numCV]), decreasing=TRUE)[topN], numCV]) %>% kable_styling(),
kable(CCA3$cc3.weight.x[order(abs(CCA3$cc3.weight.x[,numCV]), decreasing=TRUE)[topN], numCV]) %>% kable_styling(),
kable(CCA3$cc3.weight.y[order(abs(CCA3$cc3.weight.y[,numCV]), decreasing=TRUE)[topN], numCV]) %>% kable_styling()
)) 

```

## Saving results

```{r}
save.CCA(CCA3, file="CCA3")
```


# References {.unnumbered}


# Session info

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
